### PR TITLE
Interaction between Turbo and TimeAndSpace

### PR DESCRIPTION
**Pull request by blackshadowshade**
Addresses #2253.

So, I'm putting this incomplete pull request up as a place to get all our comments in the one place.

### CODE REVIEW

**Comment by blackshadowshade** [comment]
I note here that manual testing fails, since rerolling Snuff's ^p(Y)! die to an odd number doesn't give that player a new turn.

@cgolubi1, as well as putting a response test for the first turn, would you mind adding another turn or two where the player *does* get an extra turn from rerolling the ^p(Y) die to an odd value, please?

**Comment by cgolubi1** [comment]
Roger.  I don't have enough internet to pull down your branches right now, but once i do, i'll snag everything and add those turns.

**Comment by blackshadowshade** [comment]
I've rebased this to get rid of conflicts (and renumbered the responder test to 55).

**Comment by cgolubi1** [comment]
Okay, one thing at a time: here's the result data from the first turn:

```
$ git diff
diff --git a/test/src/api/responder03Test.php b/test/src/api/responder03Test.php
index a5dc7c4..f2689d1 100644
--- a/test/src/api/responder03Test.php
+++ b/test/src/api/responder03Test.php
@@ -3224,6 +3224,24 @@ class responder03Test extends responderTestFramework {
             $retval, array(array(0, 4), array(1, 0)),
             $gameId, 1, 'Power', 0, 1, '', array(4 => 10));
 
-        // FIXME: add a check of the response data to this once the bug is fixed and it's easy to get the data
+        $expData['activePlayerIdx'] = 1;
+        $expData['validAttackTypeArray'] = array('Power', 'Skill');
+        $expData['playerDataArray'][0]['waitingOnAction'] = FALSE;
+        $expData['playerDataArray'][1]['waitingOnAction'] = TRUE;
+        $expData['playerDataArray'][0]['roundScore'] = 25;
+        $expData['playerDataArray'][0]['sideScore'] = -16.7;
+        $expData['playerDataArray'][1]['roundScore'] = 50;
+        $expData['playerDataArray'][1]['sideScore'] = 16.7;
+        $expData['playerDataArray'][0]['activeDieArray'][4]['value'] = 2;
+        $expData['playerDataArray'][0]['activeDieArray'][4]['sides'] = 10;
+        $expData['playerDataArray'][0]['activeDieArray'][4]['description'] = 'TimeAndSpace Turbo X Swing Die (with 10 sides)';
+        $expData['playerDataArray'][0]['activeDieArray'][4]['properties'] = array('HasJustTurboed');
+        $expData['playerDataArray'][0]['capturedDieArray'][] = array('value' => 1, 'sides' => 20, 'recipe' => '(20)', 'properties' => array('
+        array_splice($expData['playerDataArray'][1]['activeDieArray'], 4, 1);
+        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder003', 'message' => 'responder003 per
+        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder003', 'message' => 'Turbo die ^(X=5)
+        $expData['gameActionLogCount'] += 2;
+
+        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
     }
 }
```

I'll add the additional turns soon.

**Comment by blackshadowshade** [comment]
The pasted diff appears to be truncated. I'll add the missing stuff myself.

**Comment by cgolubi1** [comment]
Whoops, sorry about the truncation.  Still working on adding the next turn.

**Comment by cgolubi1** [comment]
Yeah, looks like you still have a problem.  Unless i'm confused, responder003 should get another turn after move 4, and doesn't:

```
        ////////////////////
        // Move 03 - responder004 Power attack using
        $_SESSION = $this->mock_test_user_login('responder004');
        $this->verify_api_submitTurn(
            array(8),
            'responder004 performed Power attack using [(20):1] against [g(10):1]; Defender g(10) was captured; Attacker (20) rerolled 1 => 8. ',
            $retval, array(array(0, 0), array(1, 0)),
            $gameId, 1, 'Power', 1, 0, '');
        $_SESSION = $this->mock_test_user_login('responder003');

        $expData['activePlayerIdx'] = 0;
        $expData['validAttackTypeArray'] = array('Power', 'Skill', 'Shadow', 'Trip');
        $expData['playerDataArray'][0]['waitingOnAction'] = TRUE;
        $expData['playerDataArray'][1]['waitingOnAction'] = FALSE;
        $expData['playerDataArray'][0]['roundScore'] = 20;
        $expData['playerDataArray'][0]['sideScore'] = -26.7;
        $expData['playerDataArray'][1]['roundScore'] = 60;
        $expData['playerDataArray'][1]['sideScore'] = 26.7;
        $expData['playerDataArray'][1]['activeDieArray'][0]['value'] = 8;
        array_splice($expData['playerDataArray'][0]['activeDieArray'], 0, 1);
        $expData['playerDataArray'][0]['turboSizeArray'] = array("3" => array(4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20));
        $expData['playerDataArray'][0]['optRequestArray'] = array('2' => array(10, 20));
        $expData['playerDataArray'][0]['activeDieArray'][3]['properties'] = array();
        $expData['playerDataArray'][0]['capturedDieArray'][0]['properties'] = array();
        $expData['playerDataArray'][1]['capturedDieArray'][] = array('value' => 1, 'sides' => 10, 'recipe' => 'g(10)', 'properties' => array('WasJustCaptured'));
        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder004', 'message' => 'responder004 performed Power attack using [(20):1] against [g(10):1]; Defender g(10) was captured; Attacker (20) rerolled 1 => 8'));
        $expData['gameActionLogCount'] += 1;
        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);


        ////////////////////
        // Move 04 - responder003 Power attack using ^(X)! - gets extra turn
        $this->verify_api_submitTurn(
            array(3),
            'responder003 performed Power attack using [^(X=10)!:2] against [(20):1]; Defender (20) was captured; Attacker ^(X=10)! rerolled from 2. Turbo die ^(X=10)! changed size from 10 to 6 sides, recipe changed from ^(X=10)! to ^(X=6)!, rolled 3. ',
            $retval, array(array(0, 3), array(1, 1)),
            $gameId, 1, 'Power', 0, 1, '', array(3 => 6));

        // BUG: should still be responder003's turn
        $expData['activePlayerIdx'] = 1;
        $expData['validAttackTypeArray'] = array('Power', 'Skill');
        $expData['playerDataArray'][0]['waitingOnAction'] = FALSE;
        $expData['playerDataArray'][1]['waitingOnAction'] = TRUE;

        $expData['playerDataArray'][0]['roundScore'] = 38;
        $expData['playerDataArray'][0]['sideScore'] = -8.0;
        $expData['playerDataArray'][1]['roundScore'] = 50;
        $expData['playerDataArray'][1]['sideScore'] = 8.0;
        array_splice($expData['playerDataArray'][1]['activeDieArray'], 1, 1);
        $expData['playerDataArray'][0]['activeDieArray'][3]['value'] = 3;
        $expData['playerDataArray'][0]['activeDieArray'][3]['sides'] = 6;
        $expData['playerDataArray'][0]['activeDieArray'][3]['description'] = 'TimeAndSpace Turbo X Swing Die (with 6 sides)';
        $expData['playerDataArray'][0]['activeDieArray'][3]['properties'] = array('HasJustTurboed');
        $expData['playerDataArray'][1]['capturedDieArray'][0]['properties'] = array();
        $expData['playerDataArray'][0]['capturedDieArray'][] = array('value' => 1, 'sides' => 20, 'recipe' => '(20)', 'properties' => array('WasJustCaptured'));
        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder003', 'message' => 'responder003 performed Power attack using [^(X=10)!:2] against [(20):1]; Defender (20) was captured; Attacker ^(X=10)! rerolled from 2'));
        array_unshift($expData['gameActionLog'], array('timestamp' => 'TIMESTAMP', 'player' => 'responder003', 'message' => 'Turbo die ^(X=10)! changed size from 10 to 6 sides, recipe changed from ^(X=10)! to ^(X=6)!, rolled 3'));
        $expData['gameActionLogCount'] += 2;
        $retval = $this->verify_api_loadGameData($expData, $gameId, 10);
```

**Comment by blackshadowshade** [comment]
Right, I think that this bug is now fixed. I've rebased to one commit. If CircleCI passes, we should be able to move onto dev site testing, I think.

File: test/src/api/responder03Test.php [hunk 1]
```diff
@@ -3379,4 +3379,173 @@ public function test_interface_game_054() {
...
+     * @depends responder00Test::test_request_savePlayerInfo
+     *
+     * Regression-testing interaction between Turbo and TimeAndSpace
+     */
+    public function test_interface_game_055() {
+
+        // responder003 is the POV player, so if you need to fake
+        // login as a different player e.g. to submit an attack, always
+        // return to responder003 as soon as you've done so
+        $this->game_number = 53;
```
Comment (cgolubi1) [inline test/src/api/responder03Test.php]: To fix before merging: game 055 should have game_number 55.

**Comment by cgolubi1** [comment]
This will want replay testing as well as dev testing; let's see how many of those i can stand up this morning.

**Comment by cgolubi1** [comment]
Stood up the replay site, didn't get to the dev site yet.  Soon!

File: test/src/api/responder03Test.php (hunk 1 above)
Comment (blackshadowshade) [reply test/src/api/responder03Test.php]: Should be fixed now.

**Comment by cgolubi1** [comment]
One note is that a side effect of this change is that the log ordering has changed a bit --- now turbo size changes are reported before T&S new turns, rather than after.  I think this is totally harmless, just mentioning it because it's showing up in the replay stats:

```
2) responder99Test::test_interface_game_00089
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'responder003 performed Skill attack using [^(3):1,ftz(14):11,tz(1/30=1)!:1] against [q(20):13]; Defender q(20) was captured; Attacker ^(3) rerolled 1 => 1; Attacker ftz(14) rerolled 11 => 8; Attacker tz(1/30=1)! rerolled from 1. responder003 gets another turn because a Time and Space die rolled odd. Turbo die tz(1/30=1)! changed size from 1 to 30 sides, recipe changed from tz(1/30=1)! to tz(1/30=30)!, rolled 25. '
+'responder003 performed Skill attack using [^(3):1,ftz(14):11,tz(1/30=1)!:1] against [q(20):13]; Defender q(20) was captured; Attacker ^(3) rerolled 1 => 1; Attacker ftz(14) rerolled 11 => 8; Attacker tz(1/30=1)! rerolled from 1. Turbo die tz(1/30=1)! changed size from 1 to 30 sides, recipe changed from tz(1/30=1)! to tz(1/30=30)!, rolled 25. responder003 gets another turn because a Time and Space die rolled odd. '
```

I should probably bother to instrument the replay rig to correct for this, though, because a game like this (where both T&S and Turbo exist, but nothing should have changed) is exactly the kind of game where i want testing not to blow up on the message comparison so i can see what else is different.

**Comment by blackshadowshade** [comment]
It's actually reassuring for me to see that logging change, because the change of order makes a lot of sense---that we can't determine if the player gets another turn until after the die is rolled.

**Comment by cgolubi1** [comment]
Yeah, that makes sense.  Anyway, i stood up a dev site --- irilyth, can you sync up 54.167.124.56 to blip?

**Comment by cgolubi1** [comment]
Okay!  That took long enough.  Here's my attempt at a version of update_replay_games that will DTRT with the log reorder:

```
#!/usr/bin/python
##### update_replay_games
# This should have the same functionality as prep_replay_games, but
# can be modified to adjust the API output to account for changes
# between the master version against which games were recorded, and
# the version under test.  One-off changes which will not be needed
# once the API update is committed should be made in this file and
# should not be committed.  Support for permanent API changes should
# be made in random_ai.py.

import sys
sys.path.append('lib')

import random_ai

def is_ts_odd_msg(message):
  return message.endswith('gets another turn because a Time and Space die rolled odd')

def is_turbo_change_msg(message):
  return message.startswith('Turbo') and 'changed size from' in message

########################################################################
# Custom classes
class PHPBMClientOutputWriter(random_ai.PHPBMClientOutputWriter):
  # If you need to make any changes to the syntax of PHP output which is
  # translated from the internal game state format for regression-testing
  # previously recorded games against this PR (e.g. if the new code
  # changes the API and you need to reflect those changes for replays
  # to succeed), make those overrides here for clarity during testing.
  #
  # As a first resort, trying overriding apply_known_changes_to_game_data()
  def apply_known_changes_to_game_data(self, data):
    for swap_a, swap_b in self.swap_action_lines:
      messages = [l['message'] for l in data['gameActionLog']]
      a_idx = None
      try:
        a_idx = messages.index(swap_a)
      except ValueError:
        continue
      b_indices = [i for i in range(len(messages)) if messages[i] == swap_b]
      if a_idx + 1 in b_indices:
        data['gameActionLog'][a_idx]['message'] = swap_b
        data['gameActionLog'][a_idx + 1]['message'] = swap_a
      else:
        if a_idx == len(messages) - 1:
          data['gameActionLog'][a_idx]['message'] = swap_b
    return data

  def _write_entry_type_submitTurn(self, entry):
    message_lines = entry['retval'].message[:-2].split('. ')
    ts_odd_idx = None
    turbo_change_idx = None
    for i in range(len(message_lines)):
      if is_ts_odd_msg(message_lines[i]):
        ts_odd_idx = i
        ts_odd_msg = message_lines[i]
      elif is_turbo_change_msg(message_lines[i]):
        turbo_change_idx = i
        turbo_change_msg = message_lines[i]
    if ts_odd_idx is not None and turbo_change_idx is not None:
      message_lines[ts_odd_idx] = turbo_change_msg
      message_lines[turbo_change_idx] = ts_odd_msg
      entry['retval'].message = '. '.join(message_lines) + '. '
      self.swap_action_lines.append([turbo_change_msg, ts_odd_msg])
    random_ai.PHPBMClientOutputWriter._write_entry_type_submitTurn(self, entry)

########################################################################
# Initial setup
if len(sys.argv) < 2:
  raise ValueError("Usage: update_replay_games <dirname>")

DIRNAME = sys.argv[1]

d = PHPBMClientOutputWriter(DIRNAME)
for gamepath in d.find_games():
  sys.stderr.write("Updating and prepping game for replay from: %s\n" % gamepath)
  d.swap_action_lines = []
  d.translate_game(gamepath)
```

Rebuilding the replay site, and let's see how it goes.  Now i expect failures when we switch from getting another turn to not, or vice versa, and i expect to be able to eyeball those and see them look right.

**Comment by cgolubi1** [comment]
Alright, so here's what i see on the replay site.  Most things pass, the only failures are:
* I didn't catch the "remained the same size" turbo message case in my update_replay_games update, so those failed
* There was one log mismatch failure which looked like the ones i was seeing before, and which strikes me as very very likely to just be another corner case my logic didn't handle; i'm not going to debug it further.
* Failures because the behavior (got a new turn vs didn't) changed because of the bug fix - the test we added to the codebase should handle that
* Failures because the number of random die values was wrong, which i believe are a subset of the previous case --- if the extra turn caused a round to not end where previously it would have ended.  The expected output looks like this:

```
        $this->verify_api_submitTurn(
            array(7, 9, 5, 3, 9, 7, 13, 13, 15, 4, 7),
            'responder003 performed Power attack using [^(X=15)!:13] against [v(20):10]; Defender v(20) was captured; Attacker ^(X=15)! rerolled from 13. responder003 gets another turn because a Time and Space die rolled odd. responder003 gets another turn because a Time and Space die rolled odd. responder003 passed. Turbo die ^(X=15)! changed size from 15 to 9 sides, recipe changed from ^(X=15)! to ^(X=9)!, rolled 7. responder003 passed. End of round: responder003 won round 4 (35.5 vs. 21). responder004 won initiative for round 5. Initial die values: responder003 rolled [g(10):9, sp(12):5, t(4):3, (10/20=10):9, ^(X=11)!:7], responder004 rolled [v(20):13, v(20):13, v(20):15, v(20):4, v(20):7]. responder003 has dice which are not counted for initiative due to die skills: [g(10), t(4)]. ',
            $retval, array(array(0, 1), array(1, 0)),
            $gameId, 4, 'Power', 0, 1, '', array(1 => 9));
```

which is pretty weird and obviously buggy.  The new output seems to be:

```
'responder003 performed Power attack using [^(X=15)!:13] against [v(20):10]; Defender v(20) was captured; Attacker ^(X=15)! rerolled from 13. Turbo die ^(X=15)! changed size from 15 to 9 sides, recipe changed from ^(X=15)! to ^(X=9)!, rolled 7. responder003 gets another turn because a Time and Space die rolled odd. responder003 passed. ',
```

I'm not 100% clear on what the game state is that leads to us still being in the previous round, but it looks like it's responder004's turn and responder004 has a valid power attack of some sort, so i'm going to take it on faith that whatever the state it, it's less buggy than before, and not hold us up on trying to  introspect it.

**Comment by cgolubi1** [comment]
As far as i can tell, this is ready to rebase down to one commit and merge.

**Comment by blackshadowshade** [comment]
Okay, this has been rebased.

**Comment by cgolubi1** [comment]
Merging, finally.
